ably/realtime_conn_spec_test.go

```diff
@@ -3390,6 +3390,185 @@ func TestIdempotentACK(t *testing.T) {
+	dial, intercept := DialIntercept(ably.DialWebsocket)
+	c := app.NewRealtime(
+		ably.WithDial(dial),
+		ably.WithDefaultTokenParams(ably.TokenParams{
+			Capability: `{"foo":["subscribe"]}`,
+		}),
+		ably.WithAuthCallback(func(ctx context.Context, params ably.TokenParams) (ably.Tokener, error) {
+			respCh := make(chan authResponse, 1)
+			authRequests <- authRequest{
+				params: params,
+				resp:   respCh,
+			}
+			resp := <-respCh
+			return resp.token, resp.err
+		}),
+	)
+	defer safeclose(t, ablytest.FullRealtimeCloser(c), app)
+
+	stateChanges := make(ably.ConnStateChanges, 1)
+	off := c.Connection.OnAll(stateChanges.Receive)
+	defer off()
+	expectConnEvent := func(expected ably.ConnectionEvent) {
+		t.Helper()
+
+		var change ably.ConnectionStateChange
+		ablytest.Soon.Recv(t, &change, stateChanges, t.Fatalf)
+		if got := change.Event; expected != got {
+			t.Fatalf("expected: %v; got: %v", expected, got)
+		}
+	}
+
+	// Expect a first auth when connecting.
+	handleAuth(getToken, nil)
+	expectConnEvent(ably.ConnectionEventConnected)
+
+	t.Run("RTC8a1: Successful reauth with more capabilities", func(t *testing.T) {
+		var rg ablytest.ResultGroup
+		rg.GoAdd(func(ctx context.Context) error {
+			_, err := c.Auth.Authorize(ctx, nil)
+			return err
+		})
+
+		handleAuth(getToken, nil)
+		expectConnEvent(ably.ConnectionEventUpdate)
+		assertEquals(t, connected, c.Connection.State())
+
+		rg.Wait()
+	})
+
+	t.Run("RTC8a1: Successful reauth with more capabilities", func(t *testing.T) {
+		var rg ablytest.ResultGroup
+		rg.GoAdd(func(ctx context.Context) error {
+			_, err := c.Auth.Authorize(ctx, &ably.TokenParams{
+				Capability: `{"*":["*"]}`,
+			})
+			return err
+		})
+
+		handleAuth(getToken, nil)
+		expectConnEvent(ably.ConnectionEventUpdate)
+		assertEquals(t, connected, c.Connection.State())
+
+		rg.Wait()
+	})
+
+	t.Run("RTC8a1: Unsuccessful reauth with capabilities downgrade", func(t *testing.T) {
+		ch := c.Channels.Get("notfoo")
+		err := ch.Attach(context.Background())
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		channelChanges := make(ably.ChannelStateChanges, 1)
+		off := ch.OnAll(channelChanges.Receive)
+		defer off()
+
+		var rg ablytest.ResultGroup
+		rg.GoAdd(func(ctx context.Context) error {
+			_, err := c.Auth.Authorize(ctx, &ably.TokenParams{
+				Capability: `{"foo":["publish"]}`,
+			})
+			return err
+		})
+
+		handleAuth(getToken, nil)
+		expectConnEvent(ably.ConnectionEventUpdate)
+		assertEquals(t, connected, c.Connection.State())
+
+		rg.Wait()
+
+		var change ably.ChannelStateChange
+		ablytest.Soon.Recv(t, &change, channelChanges, t.Fatalf)
+		assertEquals(t, ably.ChannelEventFailed, change.Event)
+		assertEquals(t, chFailed, ch.State())
+	})
+
+	t.Run("RTC8a3: Authorize waits for connection update", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		connectedMsg := intercept(ctx, ably.ActionConnected)
+
+		authorizeDone := make(chan struct{})
+		go func() {
```

Comment:
open goroutine

Reply:
Yes, we wait for `authorizeDone` below, so we check there isn't a leak.

Reply:
Still, its a explicit go routine, we can follow standard `rgGroup`

Reply:
Don't really see the point but done at eb709f3.